tokio/src/sync/set_once.rs

```diff
@@ -0,0 +1,409 @@
+        } else {
+            None
+        }
+    }
+
+    /// Sets the value of the `SetOnce` to the given value if the `SetOnce` is
+    /// empty.
+    ///
+    /// If the `SetOnce` already has a value, this call will fail with an
+    /// [`SetOnceError`].
+    ///
+    /// [`SetOnceError`]: crate::sync::SetOnceError
+    pub fn set(&self, value: T) -> Result<(), SetOnceError<T>> {
+        if self.initialized() {
+            return Err(SetOnceError::from(value));
+        }
+
+        // SAFETY: lock the mutex to ensure only one caller of set
+        // can run at a time.
+        let guard = self.lock.lock();
+
+        if self.initialized() {
+            // If the value is already set, we return an error
+            drop(guard);
+
+            return Err(SetOnceError::from(value));
+        }
+
+        // SAFETY: We have locked the mutex and checked if the value is
+        // initalized or not, so we can safely write to the value
+        unsafe {
+            self.value.with_mut(|ptr| (*ptr).as_mut_ptr().write(value));
+        }
+
+        // Using release ordering so any threads that read a true from this
+        // atomic is able to read the value we just stored.
+        self.value_set.store(true, Ordering::Release);
+
+        drop(guard);
+
+        // notify the waiting wakers that the value is set
+        self.notify.notify_waiters();
+
+        Ok(())
+    }
+
+    /// Takes the value from the cell, destroying the cell in the process.
+    /// Returns `None` if the cell is empty.
+    pub fn into_inner(self) -> Option<T> {
+        // TODO: Use get_mut()
+        let value_set = self.value_set.load(Ordering::Relaxed);
+
+        if value_set {
+            // Since we have taken ownership of self, its drop implementation
+            // will be called by the end of this function, to prevent a double
+            // free we will set the value_set to false so that the drop
+            // implementation does not try to drop the value again.
+            self.value_set.store(false, Ordering::Relaxed);
+
+            // SAFETY: The SetOnce is currently initialized, we can assume the
+            // value is initialized and return that, when we return the value
+            // we give the drop handler to the return scope.
+            Some(unsafe { self.value.with_mut(|ptr| ptr::read(ptr).assume_init()) })
+        } else {
+            None
+        }
+    }
+
+    /// Waits until set is called. The future returned will keep blocking until
+    /// the `SetOnce` is initialized.
+    ///
+    /// If the `SetOnce` is already initialized, it will return the value
+    /// immediately.
+    ///
+    /// # Note
+    ///
+    /// This will keep waiting until the `SetOnce` is initialized, so it
+    /// should be used with care to avoid blocking the current task
+    /// indefinitely.
+    pub async fn wait(&self) -> &T {
+        loop {
+            let notify_fut = self.notify.notified();
```

Comment:
It could make sense to add a fast-path here. That way, we don't need to access the `Notify` at all if the future is already set.
```suggestion
    pub async fn wait(&self) -> &T {
        loop {
            if let Some(value) = self.get() {
	        return value;
	    }

            let notify_fut = self.notify.notified();
```